### PR TITLE
Add interaction graph prototype

### PR DESCRIPTION
**Pull request by kiootic**
ref #73

### CODE REVIEW

File: pkg/auth/dependency/newinteraction/graph.go [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+
+func (g *Graph) AppendingNode(n Node) *Graph {
+	nodes := make([]Node, len(g.Nodes)+1)
+	copy(nodes, g.Nodes)
+	nodes[len(nodes)-1] = n
+
+	return &Graph{
+		GraphID:    g.GraphID,
+		InstanceID: "",
+		Nodes:      nodes,
```
Comment (louischan-oursky) [inline pkg/auth/dependency/newinteraction/graph.go:46]: Copy `Intent`

File: pkg/auth/dependency/newinteraction/node_select_identity_end.go [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+func (n *NodeSelectIdentityEnd) Apply(ctx *Context, graph *Graph) error {
+	panic("implement me")
+}
+
+func (n *NodeSelectIdentityEnd) DeriveEdges(ctx *Context, graph *Graph) ([]Edge, error) {
+	edge, err := graph.Intent.(interface {
+		AfterSelectIdentity(node *NodeSelectIdentityEnd) (Edge, error)
+	}).AfterSelectIdentity(n)
+
+	return []Edge{edge}, err
```
Comment (louischan-oursky) [inline pkg/auth/dependency/newinteraction/node_select_identity_end.go:33]: I think it is better to add `DeriveEdges(ctx *Context, graph *Graph, node Node) ([]Edge, error)` to interface `Intent`. So here we just write

```golang
func (n *NodeSelectIdentityEnd) DerivesEdges(ctx *Context, graph *Graph) ([]Edge, error) {
  return graph.Intent.DerivesEdges(ctx, graph, n)
}
```

File: pkg/auth/dependency/newinteraction/intent_login.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+func (i *IntentLogin) DeriveFirstNode(ctx *Context, graph *Graph) (Node, error) {
+	spec := EdgeSelectIdentityBegin{}
+	return spec.Instantiate(ctx, graph, i)
+}
+
+func (i *IntentLogin) GetUseAnonymousUser() bool {
+	return i.UseAnonymousUser
+}
+
+func (i *IntentLogin) AfterSelectIdentity(node *NodeSelectIdentityEnd) (Edge, error) {
```
Comment (louischan-oursky) [inline pkg/auth/dependency/newinteraction/intent_login.go:16]:
```golang
func (i *IntentLogin) DeriveEdges(ctx *Context, graph *Graph, node Node) ([]Edge, error) {
  switch node.(type) {
  case *NodeSelectIdentityEnd:
    return []Edge{&EdgeAuthenticationBegin{State: AuthenticationPrimary, Identity: node.Identity}}, nil
  default:
    panic("interaction: unexpected node")
  }
}

```

**Comment by louischan-oursky** [comment]
`ExtraErrorRedirectURI` and `ExtraRedirectURI` should become first-class fields of Graph. These two fields should be provided when the graph is created.

**Comment by louischan-oursky** [comment]
Can you fix this compilation as well?

```diff
commit b598925d52ae4eecb7848e91ef53bfba6da5c37b
Author: Louis Chan <louischan@oursky.com>
Date:   Tue Jul 21 17:21:17 2020 +0800

    Fix compilation error

diff --git a/pkg/auth/dependency/newinteraction/intent.go b/pkg/auth/dependency/newinteraction/intent.go
index f5ae8857..07e34613 100644
--- a/pkg/auth/dependency/newinteraction/intent.go
+++ b/pkg/auth/dependency/newinteraction/intent.go
@@ -5,7 +5,7 @@ import (
 )
 
 type Intent interface {
-	InstantiateRootNode(ctx *Context) (Node, error)
+	InstantiateRootNode(ctx *Context, graph *Graph) (Node, error)
 	DeriveEdges(ctx *Context, graph *Graph, node Node) ([]Edge, error)
 }
 
diff --git a/pkg/auth/dependency/newinteraction/service.go b/pkg/auth/dependency/newinteraction/service.go
index 5e5e313a..6a0a613e 100644
--- a/pkg/auth/dependency/newinteraction/service.go
+++ b/pkg/auth/dependency/newinteraction/service.go
@@ -53,7 +53,7 @@ func (s *Service) WithContext(fn func(*Context) error) error {
 
 func (s *Service) NewGraph(ctx *Context, intent Intent) (*Graph, error) {
 	graph := newGraph(intent)
-	node, err := graph.Intent.InstantiateRootNode(ctx)
+	node, err := graph.Intent.InstantiateRootNode(ctx, graph)
 	if err != nil {
 		return nil, err
 	}

```

**Comment by kiootic** [comment]
@louischan-oursky Fixed. You can review & merge this first, I'll open another PR for further works.

**Comment by louischan-oursky** [comment]
Maybe merge into a `interaction-graph` branch first?

**Comment by louischan-oursky** [comment]
Created the branch `interaction-graph`. Can you switch the target branch to it?

**Pull request merged**